src/dsp_tools/commands/excel2json/new_lists.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import annotations
+
+from typing import Any
+
+import pandas as pd
+import regex
+
+from dsp_tools.commands.excel2json.models.input_error import ListNodeProblem
+from dsp_tools.commands.excel2json.models.input_error import ListSheetProblem
+from dsp_tools.commands.excel2json.models.list_node import ListNode
+from dsp_tools.commands.excel2json.models.list_node import ListRoot
+from dsp_tools.models.exceptions import InputError
+
+
+def _fill_id_and_parent_id_columns(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+    df = _fill_auto_id_column(df, preferred_language)
+    df["id"] = df["ID (optional)"].fillna(df["auto_id"])
+    df = _fill_parent_id(df, preferred_language)
+    return df
+
+
+def _fill_auto_id_column(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+    """For every node without manual ID, take the label of the preferred language as ID."""
+    df["auto_id"] = pd.NA
+    if not df["ID (optional)"].isna().any():
+        return df
+    if pd.isna(df.at[0, "ID (optional)"]):
+        df.at[0, "auto_id"] = df.at[0, f"{preferred_language}_list"]
+    columns = sorted(_get_columns_of_preferred_lang(df.columns, preferred_language), reverse=True)
+    for i, row in df.iterrows():
+        if pd.isna(row["ID (optional)"]):
+            for col in columns:
+                if pd.notna(row[col]):
+                    df.at[i, "auto_id"] = row[col]
+                    break
+    return df
+
+
+def _fill_parent_id(df: pd.DataFrame, preferred_language: str) -> pd.DataFrame:
+    """Create an extra column with the ID of the parent node."""
+    # To start, all rows get the ID of the list. These will be overwritten if the row has another parent.
+    df["parent_id"] = df.at[0, "id"]
+    columns = _get_columns_of_preferred_lang(df.columns, preferred_language)
+    for col in columns:
+        grouped = df.groupby(col)
+        for name, group in grouped:
+            if name == "nan":
+                # Leaf node: ID already assigned
+                pass
+            elif group.shape[0] > 1:
+                # The first row already has the correct ID assigned
+                rest_index = list(group.index)[1:]
+                df.loc[rest_index, "parent_id"] = group.at[group.index[0], "id"]
+    return df
+
+
+def _make_one_list(df: pd.DataFrame, sheet_name: str) -> ListRoot | ListSheetProblem:
+    node_dict, problems = _make_list_nodes(df)
+    cols = [x for x in df.columns if regex.search(r"^(en|de|fr|it|rm)_list$", x)]
+    root = ListRoot.create(
+        id_=df.at[0, "id"],
+        labels=_get_labels(df.iloc[0], cols),
+        sheet_name=sheet_name,
+    )
+    match (root, problems):
+        case (ListRoot(), list(ListNodeProblem())):
+            return ListSheetProblem(sheet_name, root_problems={}, nodes=problems)
```

Comment:
Sad that the type deduction doesn't properly work here, but then I suggest this:

```
match (root, node_problems):
        case (ListRoot(), list(ListNodeProblem())):
            return ListSheetProblem(sheet_name, root_problems={}, node_problems=node_problems)
        case (ListSheetProblem(), _):
            root = cast(ListSheetProblem, root)
            root.node_problems = node_problems
            return root
    root = cast(ListRoot, root)
    root.nodes = _add_nodes_to_parent(node_dict, df.at[0, "id"])
    return root
```

This would make the code better understandable (at least for me). The price to pay would be two manual castings.

Reply:
from what I understand, the following should get the typing right:

```python
match (root, problems):
    case (ListRoot(), []): # only matches empty list, so the happy path
        root.nodes = _add_nodes_to_parent(node_dict, df.at[0, "id"])
        return root
    case (ListRoot(), _): # non-empty, so unhappy path
        return ListSheetProblem(sheet_name, root_problems={}, nodes=problems)
    case (ListSheetProblem(), _):
        root.nodes = problems
        return root
```